### PR TITLE
Add support for exceptions/asserts in CUDA device functions

### PR DESCRIPTION
**Pull request by testhound**
This pull request adds support for raising exceptions and asserts in CUDA device functions and address this open [issue](https://github.com/numba/numba/issues/8036).

### CODE REVIEW

File: numba/cuda/dispatcher.py [hunk 1]
```diff
@@ -335,7 +359,31 @@ def launch(self, args, griddim, blockdim, stream=0, sharedmem=0):
...
+
+            gpu_mem = cuda.cudadrv.driver.MemoryPointer(self.target_context,
+                                                        excinfo_val.pickle_buf,
+                                                        pickle_buf_sz)
+
+            pickle_buf_host = np.empty((excinfo_val.pickle_bufsz,),
+                                       dtype=np.byte)
+
+            driver.device_to_host(pickle_buf_host.ctypes.data,
+                                  gpu_mem.device_ctypes_pointer,
```
Comment (sklam) [inline numba/cuda/dispatcher.py:379]:
```suggestion
                                  gpu_mem,
```

**Comment by testhound** [comment]
gpuci run tests

**Comment by testhound** [comment]
gpuci run tests

**Comment by testhound** [comment]
gpuci run tests

**Comment by testhound** [comment]
gpuci run tests

**Comment by testhound** [comment]
@gmarkall I have updated the PR and it is ready for review. Before changing the status from a draft PR, I wanted to get your feedback on the changes required due to the calling convention change. Because we are now passing a pointer to execinfo_t as a argument, any external .c or .cu program will need to be modified to include the argument.

**Comment by testhound** [comment]
gpuci run tests

**Comment by testhound** [comment]
gpuci run tests

**Comment by gmarkall** [comment]
> Because we are now passing a pointer to execinfo_t as a argument, any external .c or .cu program will need to be modified to include the argument.

How difficult is it to make it so that functions declared with `declare_device()` to use the current calling convention, in order to avoid breaking existing code?

Another question - what is the impact of performance (or just register usage, as a proxy for performance as a quick approximation)? Can you gather register usage information for code generated before this PR, and after it, and compare the two to give an idea of how much the performance overhead might be?

We may need to consider making the exception support opt-in (either by a `@cuda.jit` decorator kwarg and / or config variable).

**Comment by gmarkall** [comment]
Closing as per discussion in the dev meeting on 2023-07-04: https://github.com/numba/numba/wiki/Minutes_2023_07_04

Summary: this changes the ABI and may increase register pressure as well as breaking the interface with externally-linked code (PTX and CUDA / C++ source). Given current priorities, making exceptions work in device functions will be on hold for the forseeable future.

**Pull request closed without merging**